src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerExtensions.cs

```diff
@@ -95,7 +97,54 @@ public static IApplicationBuilder UseExceptionHandler(this IApplicationBuilder a
                 throw new ArgumentNullException(nameof(options));
             }
 
-            return app.UseMiddleware<ExceptionHandlerMiddleware>(Options.Create(options));
+            var iOptions = Options.Create(options);
+            return SetExceptionHandlerMiddleware(app, iOptions);
+        }
+
+        private static IApplicationBuilder SetExceptionHandlerMiddleware(IApplicationBuilder app, IOptions<ExceptionHandlerOptions>? options)
+        {
+            const string globalRouteBuilderKey = "__GlobalEndpointRouteBuilder";
+            // Check if UseRouting() has been called so we know if it's safe to call UseRouting()
+            // otherwise we might call UseRouting() when AddRouting() hasn't been called which would fail
+            if (app.Properties.TryGetValue("__EndpointRouteBuilder", out _) || app.Properties.TryGetValue(globalRouteBuilderKey, out _))
+            {
+                return app.Use(next =>
+                {
+                    var loggerFactory = app.ApplicationServices.GetRequiredService<ILoggerFactory>();
+                    var diagnosticListener = app.ApplicationServices.GetRequiredService<DiagnosticListener>();
+
+                    if (options is null)
+                    {
+                        options = app.ApplicationServices.GetRequiredService<IOptions<ExceptionHandlerOptions>>();
+                    }
+
+                    if (!string.IsNullOrEmpty(options.Value.ExceptionHandlingPath) && options.Value.ExceptionHandler is null)
+                    {
+                        app.Properties.TryGetValue(globalRouteBuilderKey, out var routeBuilder);
+                        // start a new middleware pipeline
+                        var builder = app.New();
+                        if (routeBuilder is not null)
+                        {
+                            // use the old routing pipeline if it exists so we preserve all the routes and matching logic
+                            builder.Properties[globalRouteBuilderKey] = routeBuilder;
```

Comment:
Isn't this redundant since we're calling `app.New()` which already copies all the properties? I could see needing this if we did `new ApplicationBuilder(app.ApplicationServices)` instead. Is that what we should be doing? What if it was "__EndpointRouteBuilder" that was set and not "__GlobalEndpointRouteBuilder"?

Reply:
It doesn't copy the global endpoint route builder

Reply:
Nice. Do we have any tests that fail if this doesn't get copied?

Reply:
It's also probably worth calling out `IApplicationBuilder.New()` doesn't copy "__GlobalEndpointRouteBuilder" in a comment (at least `WebApplication`).

Reply:
> Do we have any tests that fail if this doesn't get copied?

Yes, any of the tests that set the middleware after `UseRouting()` (with global routing set) will fail.

Reply:
All the Diagnostic tests pass if I comment this line out. Is there some test I'm not running that would fail?

Reply:
> All the Diagnostic tests pass if I comment this line out

Right, because we already assume that app.New() won't have the global route builder so we copy it out before calling app.New().